pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1664,58 +1646,47 @@ func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployme
 	}
 }
 
-// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status.
-// It returns true when cluster deployment status has been updated.
+// updatePullSecretInfo creates or updates the merged pull secret for the clusterdeployment.
+// It returns true when the merged pull secret has been created or updated.
 func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
-	var err error
-	pullSecretObjExists := true
-	existingPullSecretObj := &corev1.Secret{}
-	mergedSecretName := constants.GetMergedPullSecretName(cd)
-	err = r.Get(context.TODO(), types.NamespacedName{Name: mergedSecretName, Namespace: cd.Namespace}, existingPullSecretObj)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			cdLog.Info("Existing pull secret object not found")
-			pullSecretObjExists = false
-		} else {
-			return false, errors.Wrap(err, "Error getting pull secret from cluster deployment")
-		}
+	pullSecretSecret := &corev1.Secret{}
+	switch created, err := controllerutils.GetOrCreate(
+		context.Background(),
+		r.Client,
+		client.ObjectKey{Namespace: cd.Namespace, Name: constants.GetMergedPullSecretName(cd)},
+		pullSecretSecret,
+		func() error {
+			pullSecretSecret.Type = corev1.SecretTypeDockerConfigJson
+			pullSecretSecret.StringData = map[string]string{corev1.DockerConfigJsonKey: pullSecret}
+			return errors.Wrap(
+				controllerutil.SetControllerReference(cd, pullSecretSecret, r.scheme),
+				"error setting controller reference on new merged pull secret",
+			)
+		},
```

Comment:
IMHO I find the code relatively less readable as compared with the old code we had. I understand the motive behind it, but it is relatively harder to read.

Reply:
Can you elaborate on why you find if less readable? Is it the `GetOrCreate` function in general? Or is there something about the flow of the logic in the function that is more difficult to follow?

I am inclined to keep the change for two reasons. (1) I would like to stay consistent across the code base with using `GetOrCreate`. (2) It is more concise, eliminating 28 lines of code.

Reply:
When I start reading the function `updatePullSecretInfo` in the begining we have `GetOrCreate` which does not align with the function name. IMO the function name should be like `createOrUpdateMergePullSecret` instead of `updatePullSecretInfo`.

Second, it is not easy to understand the purpose of the anonymous function.  Also the `GetOrCreate` function too hard to understand.

Reply:
> When I start reading the function `updatePullSecretInfo` in the begining we have `GetOrCreate` which does not align with the function name. IMO the function name should be like `createOrUpdateMergePullSecret` instead of `updatePullSecretInfo`.

This PR does not change the behavior of the `updatePullSecretInfo` function. If the name is not appropriate, then that is a problem that is outside of the scope of this PR.

Reply:
> Second, it is not easy to understand the purpose of the anonymous function. Also the `GetOrCreate` function too hard to understand.

It sounds to me like your main gripe is with the `GetOrCreate` function in general, not just in this particular use of the function. My main goal with the function is to consolidate some of the logic and to provide a consistent way in which the common pattern of getting a resource or creating it if it does not exist can be written.

To be clear, given the choice between the two following alternatives, your choice would be the one that does not use `GetOrCreate`, right?

```
pullSecretSecret := &corev1.Secret{}
switch created, err := controllerutils.GetOrCreate(
	context.Background(),
	r.Client,
	client.ObjectKey{Namespace: cd.Namespace, Name: constants.GetMergedPullSecretName(cd)},
	pullSecretSecret,
	func() error {
		pullSecretSecret.Type = corev1.SecretTypeDockerConfigJson
		pullSecretSecret.StringData = map[string]string{corev1.DockerConfigJsonKey: pullSecret}
		return errors.Wrap(
			controllerutil.SetControllerReference(cd, pullSecretSecret, r.scheme),
			"error setting controller reference on new merged pull secret",
		)
	},
); {
case err != nil:
	return false, errors.Wrap(err, "error getting or creating merged pull secret secret")
case created:
	cdLog.WithField("secretName", pullSecretSecret.Name).Info("Created the merged pull secret object successfully")
	return true, nil
}
```
```
pullSecret := &corev1.Secret{}
switch err := r.Get(
    context.Background(),
    client.ObjectKey{Namespace: cd.Namespace, Name: constants.GetMergedPullSecretName(cd)},
    pullSecret,
); {
case apierrors.IsAlreadyExists(err):
    pullSecret.Namespace = cd.Namespace
    pullSecret.Name = constants.GetMergedPullSecretName(cd)
    pullSecretSecret.Type = corev1.SecretTypeDockerConfigJson
    pullSecretSecret.StringData = map[string]string{corev1.DockerConfigJsonKey: pullSecret}
    if err := r.Create(context.Background(), pullSecret); err != nil {
        return false, errors.Wrap(err, "error creating merged pull secret secret")
    } else {
        cdLog.WithField("secretName", pullSecretSecret.Name).Info("Created the merged pull secret object successfully")
        return true, nil
    }
case err != nil:
    return false, errors.Wrap(err, "error getting merged pull secret secret")
}
```

Reply:
Either is ok with me but if we're talking margins, the latter is a little more readable to me, it's a little more clear what everything is doing whereas with the former I think I'd often be going back to look at what that func was supposed to do in particular.

Reply:
OK. Since the only ones expressing opinions are both in favor of the latter option, I will change to use that one.